### ISSUE TITLE
dynamo <> autograd.Function assumes that grad_outputs have the same subclass-ness as forward outputs

### ISSUE THREAD

**Issue by bdhirsh** [pytorch/pytorch#117662]
This is currently an issue for @drisspg's float8 logic to optionally recompute casted weights in the backward ([PR](https://github.com/pytorch-labs/float8_experimental/pull/186)), which has a custom `autograd.Function` that returns a plain tensor in the [forward](https://github.com/pytorch-labs/float8_experimental/pull/186/files#diff-00f68398c8aad5a3e946cccd7211a80841da9403d6c664452a45e04101bea6d6R211), but ends up getting a tensor subclass (`Float8Tensor`) as a grad_output in the [backward](https://github.com/pytorch-labs/float8_experimental/pull/186/files#diff-00f68398c8aad5a3e946cccd7211a80841da9403d6c664452a45e04101bea6d6R214)

Dynamo's current policy is pretty reasonable, it assumes that the subclass-ness of forward outputs will match the inputs to the backward ([here](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/variables/higher_order_ops.py#L1539)), same as AOTAutograd. However, it's possible (and true in this case), for the gradients passed to the backward to be subclasses, even though the outputs of the forward were not subclasses.

I think there are a few options in this particular case:

(1) User hints: allow the user to "instruct" dynamo about the subclass-ness of grad_outputs. Bikeshedding, but some API like:
```
torch._dynamo.mark_subclass(grad_out, Float8Tensor)
```

or even type hints, could be used to have dynamo change how it decides to mock up gradients.

(2) Figure out a workaround for this case. For example, in theory if we allow_in_graph this entire custom autograd.Function, tracing it out in AOTAutograd should do the right thing. It's not immediately clear how possible this is.

**Comment by bdhirsh** [comment]
FYI: I worked around it successfully for now with allow_in_graph: https://www.internalfb.com/phabricator/paste/view/P1045306974

**Comment by drisspg** [comment]
Some offline talk we have had on this was have this be a property on the subclass that can have a default:

'grad_tensor_type = subclass' 

Then for certain ops have this op change the property of the output of the forward 'grad_tensor_type = tensor`

Reasoning: The grad class type can likely very across different ops and would be necessary to not have this apply globally

**Comment by zou3519** [comment]
(3) Turn the autograd.Function into a custom operator. 

(4) rely on compiled autograd for this

**Comment by bdhirsh** [comment]
Custom op sounds reasonable to me - with the caveat that a user writing an autograd.Function that happens to have the issue above won't know to write it as a custom op in the first place, without better documentation of the current `autograd.Function` paper cuts with compile.

**Comment by StrongerXi** [comment]
Issue scrubbing. Is this still an issue, and how relevant is it? @zou3519 @drisspg

**Comment by zou3519** [comment]
still a problem, very relevant

**Comment by kshitij12345** [comment]
Hit the same problem, is anyone working on the above [plan](https://github.com/pytorch/pytorch/issues/117662#issuecomment-3791173360)?

Repro where this can cause silent incorrectness.
```python
import torch
import torch.utils._pytree as pytree


class Sub(torch.Tensor):
    @staticmethod
    def __new__(cls, data):
        return torch.Tensor._make_wrapper_subclass(
            cls, data.shape, dtype=torch.float32, device=data.device, requires_grad=False)

    def __init__(self, data):
        self._data = data

    def __tensor_flatten__(self):
        return ["_data"], {}

    @staticmethod
    def __tensor_unflatten__(inner, meta, size, stride):
        return Sub(inner["_data"])

    @classmethod
    def __torch_dispatch__(cls, func, types, args, kwargs=None):
        args = pytree.tree_map_only(Sub, lambda t: t._data, args)
        kwargs = pytree.tree_map_only(Sub, lambda t: t._data, kwargs or {})
        return func(*args, **kwargs)


class F(torch.autograd.Function):
    @staticmethod
    def forward(ctx, x):
        return Sub(x * 2)

    @staticmethod
    def backward(ctx, grad_output):
        scale = 100.0 if isinstance(grad_output, Sub) else 1.0
        return grad_output * scale * 2


def model(x):
    out = F.apply(x)          # Sub: subclass intermediate (not a graph output)
    return (out * 1.0).sum()  # consumed by a dense op -> real cotangent is dense


def grad_of(fn, x):
    x = x.clone().detach().requires_grad_(True)
    fn(x).backward()
    return x.grad


x = torch.randn(8)
eager = grad_of(model, x)
torch._dynamo.reset()
compiled = grad_of(torch.compile(model, fullgraph=True), x)

print("eager   :", eager.sum().item())
print("compiled:", compiled.sum().item())
print("ratio   :", (compiled.sum() / eager.sum()).item(), "(should be 1.0)")
```

Output
```python
# Output:
# eager   : 16.0
# compiled: 1600.0
# ratio   : 100.0 (should be 1.0)
```
cc: @pggPL

**Comment by zou3519** [comment]
I think @williamwen42 @anijain2305 have a new plan for this
